File: Frameworks/react-web/src/engine/auth-service.ts
import type PocketBase from 'pocketbase'
import { logInfo, logWarn, logError } from './log-service.ts'

/**
 * Authentication and role-based access control using PocketBase auth.
 *
 * ODS Ethos: The framework handles all auth complexity. Builders just add
 * `"roles": ["admin"]` to their spec elements, and AuthService makes it work.
 *
 * PocketBase handles password hashing, tokens, and sessions natively.
 * ODS roles are stored as a JSON array field on the user record.
 *
 * Login uses email as the primary identifier. OAuth2 providers configured
 * in PocketBase are automatically discovered and shown on the login screen.
 */
export class AuthService {
  private pb: PocketBase
  private _isAdminSetUp = false
  private _isInitialized = false
  /** When true, the PocketBase superadmin is running this app — bypass all role checks. */
  private _isSuperAdmin = false
  /** Cached list of OAuth2 providers configured in PocketBase. */
  private _oauthProviders: OAuthProvider[] = []
  /** Rate limiting: login attempts per email. */
  private _loginAttempts: Map<string, number[]> = new Map()
  /** Session activity tracking. */
  private _lastActivity: number = Date.now()

  constructor(pb: PocketBase) {
    this.pb = pb
  }

  /** Mark that the PocketBase superadmin is operating this app. */
  setSuperAdmin(value: boolean): void {
    this._isSuperAdmin = value
  }

  get isSuperAdmin(): boolean { return this._isSuperAdmin }

  // ---------------------------------------------------------------------------
  // Public getters
  // ---------------------------------------------------------------------------

  get isInitialized(): boolean { return this._isInitialized }
  get isLoggedIn(): boolean { return this._isSuperAdmin || this.pb.authStore.isValid }
  get isGuest(): boolean { return !this.isLoggedIn }

  get currentUserId(): string | undefined {
    return this.pb.authStore.record?.id
  }

  get currentUsername(): string {
    if (this._isSuperAdmin) return 'admin'
    return (this.pb.authStore.record?.['username'] as string) ?? 'guest'
  }

  get currentDisplayName(): string {
    if (this._isSuperAdmin) return 'Admin'
    return (this.pb.authStore.record?.['displayName'] as string)
      ?? (this.pb.authStore.record?.['name'] as string)
      ?? this.currentEmail
      ?? this.currentUsername
  }

  get currentEmail(): string {
    if (this._isSuperAdmin) {
      return (this.pb.authStore.record?.['email'] as string) ?? ''
    }
    return (this.pb.authStore.record?.['email'] as string) ?? ''
  }

  get currentRoles(): string[] {
    if (this._isSuperAdmin) return ['admin', 'user']
    if (this.isGuest) return ['guest']
    const roles = this.pb.authStore.record?.['roles']
    let parsed: unknown[]
    if (Array.isArray(roles)) {
      parsed = roles
    } else if (typeof roles === 'string') {
      try { parsed = JSON.parse(roles) } catch { return ['user'] }
      if (!Array.isArray(parsed)) return ['user']
    } else {
      return ['user']
    }
    // Validate: only allow strings, normalize to lowercase.
    return parsed
      .filter((r): r is string => typeof r === 'string')
      .map(r => r.toLowerCase())
  }

  get isAdmin(): boolean {
    return this.currentRoles.includes('admin')
  }

  get isAdminSetUp(): boolean {
    return this._isAdminSetUp
  }

  /** Returns the list of OAuth2 providers configured in PocketBase. */
  get oauthProviders(): OAuthProvider[] {
    return this._oauthProviders
  }

  // ---------------------------------------------------------------------------
  // Core permission check
  // ---------------------------------------------------------------------------

  hasAccess(requiredRoles: string[] | undefined): boolean {
    if (!requiredRoles || requiredRoles.length === 0) return true
    if (this._isSuperAdmin) return true
    if (this.isAdmin) return true
    const normalized = requiredRoles.map(r => r.toLowerCase())
    return this.currentRoles.some(r => normalized.includes(r))
  }

  // ---------------------------------------------------------------------------
  // Initialization
  // ---------------------------------------------------------------------------

  /**
   * Initializes the auth service: discovers OAuth2 providers,
   * checks if an admin user exists.
   */
  async initialize(): Promise<void> {
    // Check if any user with admin role exists.
    // The roles filter may 400 if the field doesn't exist yet — fall back
    // to listing all users and checking manually.
    try {
      let foundAdmin = false
      try {
        const admins = await this.pb.collection('users').getFullList({
          filter: 'roles ~ "admin"',
          requestKey: null,
        })
        foundAdmin = admins.length > 0
      } catch {
        // roles field may not exist — check all users manually
        try {
          const allUsers = await this.pb.collection('users').getFullList({ requestKey: null })
          foundAdmin = allUsers.some((u) => {
            const roles = u['roles']
            if (typeof roles === 'string') return roles.includes('admin')
            if (Array.isArray(roles)) return (roles as string[]).includes('admin')
            return false
          })
        } catch {
          // Users collection may not exist
        }
      }
      this._isAdminSetUp = foundAdmin
    } catch {
      this._isAdminSetUp = false
    }

    // Discover OAuth2 providers configured in PocketBase
    try {
      const methods = await this.pb.collection('users').listAuthMethods({ requestKey: null } as Record<string, unknown>)
      this._oauthProviders = (methods.oauth2?.providers ?? []).map((p) => ({
        name: p.name as string,
        displayName: p.displayName as string ?? p.name as string,
        state: p.state as string,
        authURL: p.authURL as string,
        codeVerifier: p.codeVerifier as string,
        codeChallenge: p.codeChallenge as string,
        codeChallengeMethod: p.codeChallengeMethod as string,
      }))
    } catch {
      this._oauthProviders = []
    }

    this._isInitialized = true
  }

  /**
   * Ensures the PocketBase `users` auth collection exists with the
   * ODS-required extension fields (`username`, `displayName`, `roles`).
   *
   * PocketBase's `superuser upsert` CLI only creates `_superusers`, not
   * `users` — so a fresh install has no place for the sign-up /
   * setup-admin flows to write records. The admin session in this client
   * is expected to have superadmin rights when this is called (the admin
   * UI calls it right after login), so the collections API is available.
   *
   * Safe to call repeatedly; no-op if the collection already exists and
   * is usable.
   */
  async ensureUsersCollection(): Promise<void> {
    try {
      // Probe: if the collection exists and responds, we're done.
      await this.pb.collection('users').getList(1, 1, { requestKey: null })
      return
    } catch {
      // Either missing or broken — fall through to create.
    }

    try {
      await this.pb.collections.create({
        name: 'users',
        type: 'auth',
        // Custom fields on top of PB's built-in auth fields (email,
        // password, tokenKey, verified). The built-ins are added by PB
        // automatically when type === 'auth'.
        fields: [
          { name: 'username', type: 'text', required: false },
          { name: 'displayName', type: 'text', required: false },
          { name: 'roles', type: 'json', required: false, maxSize: 2000 },
        ],
        // Rules mirror ODS's "framework handles RBAC at the application
        // layer" posture. Anonymous signup allowed; only the user or a
        // superadmin can mutate or see individual records.
        listRule: 'id != ""',
        viewRule: 'id != ""',
        createRule: '',
        updateRule: 'id = @request.auth.id',
        deleteRule: null,
      } as unknown as Record<string, unknown>)
      logInfo('AuthService', 'Created users collection')
    } catch (e) {
      // Likely a benign race — another tab or the app's initialize() may
      // have created it first. Re-probe; if still unusable, log.
      try {
        await this.pb.collection('users').getList(1, 1, { requestKey: null })
      } catch {
        logError('AuthService', 'Failed to create users collection', e)
      }
    }
  }

  // ---------------------------------------------------------------------------
  // Authentication operations
  // ---------------------------------------------------------------------------

  /** Attempt to log in with email + password. Returns true on success. */
  async login(email: string, password: string): Promise<boolean> {
    // Rate limiting: max 5 attempts per email in 5 minutes.
    const now = Date.now()
    const fiveMinutes = 5 * 60 * 1000
    const key = email.toLowerCase()
    const attempts = (this._loginAttempts.get(key) ?? []).filter(t => now - t < fiveMinutes)

    if (attempts.length >= 5) {
      console.info('[SECURITY] Rate limit triggered:', key)
      throw new Error('Too many attempts. Try again later.')
    }

    try {
      await this.pb.collection('users').authWithPassword(email, password)
      // Clear attempts on success.
      this._loginAttempts.delete(key)
      console.info('[SECURITY] Login success:', key)
      return true
    } catch {
      // Record failed attempt.
      attempts.push(now)
      this._loginAttempts.set(key, attempts)
      console.info('[SECURITY] Login failure:', key)
      return false
    }
  }

  /**
   * Start OAuth2 redirect flow. Saves state to localStorage and redirects
   * the browser to the provider's auth URL. After the user authenticates,
   * PocketBase redirects back to our app where completeOAuth2() finishes
   * the flow.
   */
  async startOAuth2Redirect(providerName: string): Promise<void> {
    try {
      const methods = await this.pb.collection('users').listAuthMethods({ requestKey: null } as Record<string, unknown>)
      const provider = (methods.oauth2?.providers ?? []).find(
        (p) => p.name === providerName,
      )
      if (!provider) {
        logWarn('AuthService', `OAuth2 provider "${providerName}" not found`)
        return
      }

      // Save provider state for the callback (sessionStorage for security)
      sessionStorage.setItem('ods_oauth2_provider', providerName)
      sessionStorage.setItem('ods_oauth2_state', provider.state as string)
      sessionStorage.setItem('ods_oauth2_codeVerifier', provider.codeVerifier as string)
      sessionStorage.setItem('ods_oauth2_returnUrl', window.location.href)

      // Redirect to provider — use our app's callback URL (not PB's built-in one)
      const redirectUrl = `${window.location.origin}/oauth2-callback`
      const authUrl = (provider.authURL as string) + encodeURIComponent(redirectUrl)

      window.location.href = authUrl
    } catch (e) {
      logError('AuthService', 'OAuth2 redirect failed', e)
    }
  }

  /**
   * Complete the OAuth2 flow after the provider redirects back.
   * Extracts the code from URL params and exchanges it for a PB auth token.
   * Returns true on success.
   */
  async completeOAuth2(code: string, state: string): Promise<boolean> {
    try {
      const savedProvider = sessionStorage.getItem('ods_oauth2_provider') ?? ''
      const savedState = sessionStorage.getItem('ods_oauth2_state') ?? ''
      const codeVerifier = sessionStorage.getItem('ods_oauth2_codeVerifier') ?? ''

      if (state !== savedState) {
        logWarn('AuthService', 'OAuth2 state mismatch')
        return false
      }

      const redirectUrl = `${window.location.origin}/oauth2-callback`

      const result = await this.pb.collection('users').authWithOAuth2Code(
        savedProvider,
        code,
        codeVerifier,
        redirectUrl,
        {},  // createData
        {},  // body
        { requestKey: null },  // prevent auto-cancellation
      )

      // Ensure the OAuth user has roles set (new users won't have them)
      const roles = result.record['roles']
      if (!roles || roles === '[]' || roles === '') {
        await this.pb.collection('users').update(result.record.id, {
          roles: JSON.stringify(['user']),
        })
        await this.pb.collection('users').authRefresh()
      }

      // Clean up sessionStorage
      sessionStorage.removeItem('ods_oauth2_provider')
      sessionStorage.removeItem('ods_oauth2_state')
      sessionStorage.removeItem('ods_oauth2_codeVerifier')

      return true
    } catch (e) {
      logError('AuthService', 'OAuth2 code exchange failed', e)
      return false
    }
  }

  /** Get the saved return URL after OAuth2 callback. Validates same-origin. */
  static getOAuth2ReturnUrl(): string | null {
    const url = sessionStorage.getItem('ods_oauth2_returnUrl')
    sessionStorage.removeItem('ods_oauth2_returnUrl')
    if (!url) return null
    // Validate that the return URL has the same origin before redirecting.
    try {
      if (new URL(url).origin !== window.location.origin) {
        console.info('[SECURITY] OAuth2 return URL origin mismatch, ignoring:', url)
        return null
      }
    } catch {
      return null
    }
    return url
  }

  /** Record user activity for session timeout tracking. */
  recordActivity(): void {
    this._lastActivity = Date.now()
  }

  /** Returns true if the session has been idle for more than 30 minutes. */
  isSessionExpired(): boolean {
    const thirtyMinutes = 30 * 60 * 1000
    return Date.now() - this._lastActivity > thirtyMinutes
  }

  /** Log out the current user. */
  logout(): void {
    this.pb.authStore.clear()
  }

  /**
   * Create the initial admin account. Called from the admin setup wizard.
   * Creates a PocketBase user with admin + user roles.
   */
  async setupAdmin(email: string, password: string, displayName?: string): Promise<boolean> {
    try {
      // Reject if email matches PocketBase superadmin — they are separate identity stores
      const pbAdminEmail = (this.pb.authStore.record?.['email'] as string) ?? ''
      if (pbAdminEmail && email.toLowerCase() === pbAdminEmail.toLowerCase()) {
        logWarn('AuthService', 'Cannot create app user with same email as PocketBase superadmin')
        return false
      }

      // Generate a username from the email (before @)
      const username = email.split('@')[0].replace(/[^\w]/g, '_').toLowerCase()

      await this.pb.collection('users').create({
        username,
        password,
        passwordConfirm: password,
        email,
        displayName: displayName ?? username,
        roles: JSON.stringify(['admin', 'user']),
      })

      // Auto-login as the new admin
      await this.pb.collection('users').authWithPassword(email, password)
      this._isAdminSetUp = true
      console.info('[SECURITY] Admin setup success:', email)
      return true
    } catch (e) {
      logError('AuthService', 'Admin setup failed', e)
      console.info('[SECURITY] Admin setup failure:', email)
      return false
    }
  }

  /** Register a new user with the given role. Returns user ID on success. */
  async registerUser(params: {
    email: string
    password: string
    role: string
    displayName?: string
  }): Promise<string | null> {
    try {
      // Reject if email matches PocketBase superadmin
      const pbAdminEmail = (this.pb.authStore.record?.['email'] as string) ?? ''
      if (pbAdminEmail && params.email.toLowerCase() === pbAdminEmail.toLowerCase()) {
        logWarn('AuthService', 'Cannot register with PocketBase superadmin email')
        return null
      }

      const roles = [params.role]
      if (params.role !== 'user' && params.role !== 'guest') {
        roles.push('user')
      }

      // Generate username from email
      const username = params.email.split('@')[0].replace(/[^\w]/g, '_').toLowerCase()
        + '_' + Math.random().toString(36).slice(2, 6)

      const record = await this.pb.collection('users').create({
        username,
        password: params.password,
        passwordConfirm: params.password,
        email: params.email,
        displayName: params.displayName ?? params.email.split('@')[0],
        roles: JSON.stringify(roles),
      })

      console.info('[SECURITY] User registration success:', params.email)
      return record.id
    } catch (e) {
      logError('AuthService', 'Registration failed', e)
      console.info('[SECURITY] User registration failure:', params.email)
      return null
    }
  }

  /**
   * Validate a password against PocketBase requirements.
   * Returns null if valid, or an error message string.
   */
  static validatePassword(password: string): string | null {
    if (password.length < 8) return 'Password must be at least 8 characters.'
    if (password.length > 72) return 'Password must be 72 characters or fewer.'
    return null
  }

  /** Change password for a user. */
  async changePassword(userId: string, newPassword: string): Promise<boolean> {
    try {
      await this.pb.collection('users').update(userId, {
        password: newPassword,
        passwordConfirm: newPassword,
      })
      return true
    } catch {
      return false
    }
  }

  /** List all users (admin operation). */
  async listUsers(): Promise<Record<string, unknown>[]> {
    try {
      const records = await this.pb.collection('users').getFullList({ sort: 'created', requestKey: null })
      return records.map(r => ({
        _id: r.id,
        username: r['username'],
        email: r['email'] ?? '',
        displayName: r['displayName'] ?? r['name'] ?? r['email'] ?? r['username'],
        roles: (() => {
          const roles = r['roles']
          if (Array.isArray(roles)) return roles
          if (typeof roles === 'string') {
            try { return JSON.parse(roles) } catch { return [] }
          }
          return []
        })(),
        _createdAt: r.created,
      }))
    } catch {
      return []
    }
  }

  /** Delete a user by ID. */
  async deleteUser(userId: string): Promise<void> {
    await this.pb.collection('users').delete(userId)
  }

  /** Assign a role to a user. */
  async assignRole(userId: string, role: string): Promise<void> {
    const user = await this.pb.collection('users').getOne(userId)
    let roles: string[] = []
    try { roles = JSON.parse(user['roles'] as string) } catch { /* empty */ }
    if (!roles.includes(role)) {
      roles.push(role)
      await this.pb.collection('users').update(userId, {
        roles: JSON.stringify(roles),
      })
    }
  }

  /** Remove a role from a user. */
  async removeRole(userId: string, role: string): Promise<void> {
    const user = await this.pb.collection('users').getOne(userId)
    let roles: string[] = []
    try { roles = JSON.parse(user['roles'] as string) } catch { /* empty */ }
    roles = roles.filter(r => r !== role)
    await this.pb.collection('users').update(userId, {
      roles: JSON.stringify(roles),
    })
  }

  /** Reset to initial state. */
  reset(): void {
    this.pb.authStore.clear()
    this._isAdminSetUp = false
    this._isInitialized = false
    this._isSuperAdmin = false
    this._oauthProviders = []
    this._loginAttempts.clear()
    this._lastActivity = Date.now()
  }
}

/** Describes an OAuth2 provider discovered from PocketBase. */
export interface OAuthProvider {
  name: string
  displayName: string
  state: string
  authURL: string
  codeVerifier: string
  codeChallenge: string
  codeChallengeMethod: string
}

File: Frameworks/flutter-local/lib/engine/auth_service.dart
import 'package:flutter/foundation.dart';

import 'data_store.dart';
import 'log_service.dart';
import 'password_hasher.dart';

/// Manages authentication state and role-based access control for ODS apps.
///
/// ODS Ethos: The framework handles all auth complexity. Builders just add
/// `"roles": ["admin"]` to their spec elements, and AuthService makes it work.
///
/// This service is owned by AppEngine (like DataStore) and provides:
///   - Login/logout session management
///   - Admin setup wizard state
///   - Role-based access checks (the core `hasAccess` method)
///   - User CRUD operations (delegated to DataStore)
class AuthService extends ChangeNotifier {
  final DataStore _dataStore;

  // Session state
  String? _currentUserId;
  String? _currentUsername;
  String? _currentEmail;
  String? _currentDisplayName;
  List<String> _currentRoles = [];
  bool _isAdminSetUp = false;
  bool _isInitialized = false;

  // Rate limiting: track failed login attempts per email.
  static const int _maxFailedAttempts = 5;
  static const Duration _lockoutWindow = Duration(minutes: 5);
  final Map<String, List<DateTime>> _failedAttempts = {};

  // Session timeout tracking.
  DateTime? _lastActivity;
  static const Duration _sessionTimeout = Duration(minutes: 30);

  AuthService(this._dataStore);

  /// Inject framework-level auth state so per-app auth checks (hasAccess,
  /// isAdmin, etc.) work without a separate per-app login.
  void injectFrameworkAuth({
    required String username,
    String email = '',
    required String displayName,
    required List<String> roles,
  }) {
    _currentUserId = 'framework'; // Sentinel: signals "logged in" without a real per-app user
    _currentUsername = username.isNotEmpty ? username : 'user';
    _currentEmail = email;
    _currentDisplayName = displayName.isNotEmpty ? displayName : username.isNotEmpty ? username : 'User';
    _currentRoles = roles.isNotEmpty ? roles : const ['user'];
    _isAdminSetUp = true;
    _isInitialized = true;
  }

  // ---------------------------------------------------------------------------
  // Public getters
  // ---------------------------------------------------------------------------

  bool get isInitialized => _isInitialized;
  bool get isLoggedIn => _currentUserId != null;
  bool get isGuest => _currentUserId == null;
  bool get isAdmin =>
      _currentRoles.any((r) => r.toLowerCase() == 'admin');
  bool get isAdminSetUp => _isAdminSetUp;

  String? get currentUserId => _currentUserId;
  String get currentUsername => _currentUsername ?? 'guest';
  String get currentEmail => _currentEmail ?? '';
  String get currentDisplayName => _currentDisplayName ?? 'Guest';

  /// Returns the current user's roles. Guests get ['guest'].
  List<String> get currentRoles => isGuest ? const ['guest'] : _currentRoles;

  // ---------------------------------------------------------------------------
  // Core permission check
  // ---------------------------------------------------------------------------

  /// Checks whether the current user has access to an element with the given
  /// role restriction.
  ///
  /// Returns true when:
  ///   - [requiredRoles] is null or empty (no restriction)
  ///   - The current user is an admin (admin bypasses all restrictions)
  ///   - The current user has at least one matching role
  bool hasAccess(List<String>? requiredRoles) {
    if (requiredRoles == null || requiredRoles.isEmpty) return true;
    if (isAdmin) return true;
    final normalized = requiredRoles.map((r) => r.toLowerCase()).toList();
    return currentRoles.any((r) => normalized.contains(r.toLowerCase()));
  }

  // ---------------------------------------------------------------------------
  // Initialization
  // ---------------------------------------------------------------------------

  /// Initializes the auth service: creates auth tables and checks admin state.
  /// Called by AppEngine.loadSpec() when auth.multiUser is true.
  Future<void> initialize() async {
    await _dataStore.ensureAuthTables();
    _isAdminSetUp = await _dataStore.hasAdminUser();
    _isInitialized = true;
    notifyListeners();
  }

  // ---------------------------------------------------------------------------
  // Authentication operations
  // ---------------------------------------------------------------------------

  /// Checks whether the given email is currently rate-limited.
  /// Returns the number of minutes remaining if locked out, or 0 if not.
  int _checkRateLimit(String email) {
    final key = email.toLowerCase();
    final attempts = _failedAttempts[key];
    if (attempts == null) return 0;

    // Remove attempts outside the lockout window.
    final cutoff = DateTime.now().subtract(_lockoutWindow);
    attempts.removeWhere((t) => t.isBefore(cutoff));

    if (attempts.length >= _maxFailedAttempts) {
      final oldestRelevant = attempts.first;
      final unlockTime = oldestRelevant.add(_lockoutWindow);
      final remaining = unlockTime.difference(DateTime.now()).inMinutes + 1;
      return remaining > 0 ? remaining : 0;
    }
    return 0;
  }

  /// Records a failed login attempt for rate limiting.
  void _recordFailedAttempt(String email) {
    final key = email.toLowerCase();
    _failedAttempts.putIfAbsent(key, () => []).add(DateTime.now());
  }

  /// Clears failed login attempts after a successful login.
  void _clearFailedAttempts(String email) {
    _failedAttempts.remove(email.toLowerCase());
  }

  /// Attempts to log in with the given credentials.
  /// Returns a [LoginResult] indicating success or the reason for failure.
  Future<LoginResult> login(String email, String password) async {
    // Rate limit check.
    final lockoutMinutes = _checkRateLimit(email);
    if (lockoutMinutes > 0) {
      logWarn('AuthService', '[SECURITY] login_rate_limited: $email');
      return LoginResult(
        success: false,
        error: 'Too many failed attempts. Try again in $lockoutMinutes minute${lockoutMinutes == 1 ? '' : 's'}.',
      );
    }

    final user = await _dataStore.getUserByEmail(email);
    if (user == null) {
      _recordFailedAttempt(email);
      logInfo('AuthService', '[SECURITY] login_failed: $email (user not found)');
      return const LoginResult(success: false);
    }

    final storedHash = user['password_hash'] as String;
    final salt = user['salt'] as String;

    if (!PasswordHasher.verify(password, salt, storedHash)) {
      _recordFailedAttempt(email);
      logInfo('AuthService', '[SECURITY] login_failed: $email (bad password)');
      return const LoginResult(success: false);
    }

    _clearFailedAttempts(email);
    _currentUserId = user['_id'] as String;
    _currentUsername = user['username'] as String? ?? email;
    _currentEmail = user['email'] as String? ?? email;
    _currentDisplayName = user['display_name'] as String?;
    _currentRoles = await _dataStore.getUserRoles(_currentUserId!);
    _lastActivity = DateTime.now();
    logInfo('AuthService', '[SECURITY] login_success: $email');
    notifyListeners();
    return const LoginResult(success: true);
  }

  /// Checks whether the current session has timed out due to inactivity.
  /// Returns true if the session has been idle for more than 30 minutes.
  bool checkSessionTimeout() {
    if (!isLoggedIn || _lastActivity == null) return false;
    return DateTime.now().difference(_lastActivity!) > _sessionTimeout;
  }

  /// Records user activity to reset the session timeout timer.
  /// Call this on meaningful user interactions (navigation, form submission, etc.).
  void recordActivity() {
    if (isLoggedIn) {
      _lastActivity = DateTime.now();
    }
  }

  /// Logs out the current user, reverting to guest state.
  void logout() {
    logInfo('AuthService', '[SECURITY] logout: ${_currentUsername ?? 'unknown'}');
    _currentUserId = null;
    _currentUsername = null;
    _currentEmail = null;
    _currentDisplayName = null;
    _currentRoles = [];
    _lastActivity = null;
    notifyListeners();
  }

  /// Creates the initial admin account. Called from the admin setup wizard.
  /// Returns true on success.
  ///
  /// Pass [displayName] to store a human-readable name separate from the
  /// login email. Defaults to the email's local part when omitted.
  Future<bool> setupAdmin(
    String email,
    String password, {
    String? displayName,
  }) async {
    try {
      final salt = PasswordHasher.generateSalt();
      final hash = PasswordHasher.hash(password, salt);
      final resolvedName = (displayName != null && displayName.trim().isNotEmpty)
          ? displayName.trim()
          : email.split('@').first;

      final userId = await _dataStore.createUser(
        email: email,
        passwordHash: hash,
        salt: salt,
        displayName: resolvedName,
      );

      await _dataStore.assignRole(userId, 'admin');
      await _dataStore.assignRole(userId, 'user');

      _isAdminSetUp = true;

      // Auto-login as the new admin.
      _currentUserId = userId;
      _currentUsername = email;
      _currentEmail = email;
      _currentDisplayName = resolvedName;
      _currentRoles = ['admin', 'user'];
      _lastActivity = DateTime.now();
      logInfo('AuthService', '[SECURITY] admin_setup: $email (id=$userId)');
      notifyListeners();
      return true;
    } catch (e) {
      logError('AuthService', '[SECURITY] admin_setup_failed', e);
      return false;
    }
  }

  /// Registers a new user with the given role.
  /// Returns the user ID on success, null on failure.
  Future<String?> registerUser({
    required String email,
    required String password,
    required String role,
    String? username,
    String? displayName,
  }) async {
    try {
      final salt = PasswordHasher.generateSalt();
      final hash = PasswordHasher.hash(password, salt);

      final userId = await _dataStore.createUser(
        email: email,
        passwordHash: hash,
        salt: salt,
        username: username,
        displayName: displayName ?? username ?? email,
      );

      await _dataStore.assignRole(userId, role);
      // All non-guest users also get the 'user' base role.
      if (role != 'user' && role != 'guest') {
        await _dataStore.assignRole(userId, 'user');
      }

      logInfo('AuthService', '[SECURITY] user_created: $email role=$role (id=$userId)');
      notifyListeners();
      return userId;
    } catch (e) {
      logError('AuthService', '[SECURITY] user_creation_failed: $email', e);
      return null;
    }
  }

  /// Changes the password for a user.
  Future<bool> changePassword(String userId, String newPassword) async {
    try {
      final salt = PasswordHasher.generateSalt();
      final hash = PasswordHasher.hash(newPassword, salt);
      await _dataStore.updateUserPassword(userId, hash, salt);
      return true;
    } catch (e) {
      logError('AuthService', 'Password change failed', e);
      return false;
    }
  }

  /// Returns all users with their roles (admin-only operation).
  Future<List<Map<String, dynamic>>> listUsers() async {
    return _dataStore.listUsers();
  }

  /// Deletes a user by ID.
  Future<void> deleteUser(String userId) async {
    await _dataStore.deleteUser(userId);
    notifyListeners();
  }

  /// Assigns a role to a user.
  Future<void> assignRole(String userId, String role) async {
    await _dataStore.assignRole(userId, role);
    logInfo('AuthService', '[SECURITY] role_assigned: user=$userId role=$role');
    // Refresh current user's roles if they were affected.
    if (userId == _currentUserId) {
      _currentRoles = await _dataStore.getUserRoles(userId);
    }
    notifyListeners();
  }

  /// Removes a role from a user.
  Future<void> removeRole(String userId, String role) async {
    await _dataStore.removeRole(userId, role);
    logInfo('AuthService', '[SECURITY] role_removed: user=$userId role=$role');
    if (userId == _currentUserId) {
      _currentRoles = await _dataStore.getUserRoles(userId);
    }
    notifyListeners();
  }

  /// Resets the auth service to its initial state. Called on app close.
  void reset() {
    _currentUserId = null;
    _currentUsername = null;
    _currentEmail = null;
    _currentDisplayName = null;
    _currentRoles = [];
    _isAdminSetUp = false;
    _isInitialized = false;
    _lastActivity = null;
    _failedAttempts.clear();
  }
}

/// Result of a login attempt.
class LoginResult {
  final bool success;
  final String? error;

  const LoginResult({required this.success, this.error});
}
